File: gamepack/Decompiled/LinkDeque.java
public class LinkDeque {
   Link af = new Link();
   Link an;

   public LinkDeque() {
      this.af.ac = this.af;
      this.af.au = this.af;
   }

   public void af(Link var1) {
      if (var1.au != null) {
         var1.remove();
      }

      var1.au = this.af.au;
      var1.ac = this.af;
      var1.au.ac = var1;
      var1.ac.au = var1;
   }

   public Link an() {
      Link var1 = this.af.ac;
      if (var1 == this.af) {
         this.an = null;
         return null;
      } else {
         this.an = var1.ac;
         return var1;
      }
   }

   public Link aw() {
      Link var1 = this.an;
      if (var1 == this.af) {
         this.an = null;
         return null;
      } else {
         this.an = var1.ac;
         return var1;
      }
   }
}

File: updater-runestar/src/main/java/org/runestar/client/updater/mapper/identifiers/classes/LinkDeque.kt
package org.runestar.client.updater.mapper.identifiers.classes

import org.runestar.client.updater.mapper.abstractclasses.IdentityMapper
import org.runestar.client.updater.mapper.annotations.DependsOn
import org.runestar.client.updater.mapper.predicateutilities.and
import org.runestar.client.updater.mapper.predicateutilities.predicateOf
import org.runestar.client.updater.mapper.wrappers.Class2

@DependsOn(Link::class)
class LinkDeque : IdentityMapper.Class() {
    override val predicate = predicateOf<Class2> { it.interfaces.isEmpty() }
            .and { it != klass<Link>() }
            .and { it.instanceFields.size == 2 }
            .and { it.instanceFields.all { it.type == type<Link>() } }
}